torch/distributions/constraints.py

```diff
@@ -185,21 +188,46 @@ class _LowerTriangular(Constraint):
     Constrain to lower-triangular square matrices.
     """
     def check(self, value):
-        return (torch.tril(value) == value).min(-1)[0].min(-1)[0]
+        value_tril = batch_tril(value)
+        return (value_tril == value).view(value.shape[:-2] + (-1,)).min(-1)[0]
 
 
 class _LowerCholesky(Constraint):
     """
     Constrain to lower-triangular square matrices with positive diagonals.
     """
     def check(self, value):
+        value_tril = batch_tril(value)
+        lower_triangular = (value_tril == value).view(value.shape[:-2] + (-1,)).min(-1)[0]
+
         n = value.size(-1)
         diag_mask = torch.eye(n, n, out=value.new(n, n))
-        lower_triangular = (torch.tril(value) == value).min(-1)[0].min(-1)[0]
         positive_diagonal = (value * diag_mask > (diag_mask - 1)).min(-1)[0].min(-1)[0]
         return lower_triangular & positive_diagonal
 
 
+class _PositiveDefinite(Constraint):
+    """
+    Constrain to positive-definite matrices.
+    """
+    def check(self, value):
+        matrix_shape = value.shape[-2:]
+        batch_shape = value.unsqueeze(0).shape[:-2]
+        # TODO: replace with batched linear algebra routine when one becomes available
+        # note that `symeig()` returns eigenvalues in ascending order
+        flattened_value = value.contiguous().view((-1,) + matrix_shape)
+        return torch.stack([v.symeig()[0][:1] > 0.0 for v in flattened_value]).view(batch_shape)
```

Comment:
Is it safe to keep the greater than 0.0 check as:
```python
not (v.symeig()[0][:1]).abs() < 1e-12
```
?

Reply:
This is an [example](https://drive.google.com/file/d/1NyyMX7YJz4nJlNxOzyrO6mAZVwB2gVwv/view?usp=sharing) (dumped using `numpy.savetxt`, so you could load using `numpy.loadtxt` or `numpy.genfromtxt`). These are the top 10 eigenvalues for that matrix:
```python
9.96764812e-01 +0.00000000e+00j,
1.42575611e-10 +0.00000000e+00j,
-1.05092455e-10 +8.56060079e-11j,
-1.05092455e-10 -8.56060079e-11j,
-8.01922082e-11 +5.76392912e-11j,
-8.01922082e-11 -5.76392912e-11j,
9.43347083e-11 +1.54831723e-11j,
9.43347083e-11 -1.54831723e-11j,
7.89421910e-11 +1.71497236e-11j,
7.89421910e-11 -1.71497236e-11j
```

Reply:
That matrix is not actually symmetric, so it should also fail this constraint check anyway. Probably we should add an explicit check that the matrix is symmetric as well — I hadn't thought of that (I had figured, incorrectly and without checking, that `torch.symeig` would barf if given an assymmetric matrix).

Reply:
Do you guys want to still work on this, or should I just go ahead and merge this PR?